roboteam_ai/src/coach/OffensiveCoach.cpp

```diff
@@ -159,42 +151,105 @@ std::vector<Vector2> OffensiveCoach::getOffensivePositions(int numberOfRobots) {
-        }
-        else if (leftPointInSegment) {
-            return leftPoint;
+    if (openSegments.empty()) return world::field->get_their_goal_center();
+    auto bestSegment = getLongestSegment(openSegments);
+
+    // make two aim points which are in the corners.
+    std::pair<Vector2, Vector2> aimPoints = getAimPoints(fromPoint);
+    auto leftPoint = aimPoints.first;
+    auto rightPoint = aimPoints.second;
+
+    // check if the left and right points are in the largest segment
+    double maxY = std::max(bestSegment.first.y, bestSegment.second.y);
+    double minY = std::min(bestSegment.first.y, bestSegment.second.y);
+    bool leftPointInSegment = leftPoint.y < maxY && leftPoint.y > minY;
+    bool rightPointInSegment = rightPoint.y < maxY && rightPoint.y > minY;
+
+    // if we can aim on only one of the points, aim there, otherwise we want to aim for the centre of the largest open segment
+    if (leftPointInSegment && rightPointInSegment) {
+        // open goal (mostly), so just shoot in the middle of the largest open segment
+        return (bestSegment.first + bestSegment.second)*0.5;
+    }
+    else if (leftPointInSegment) {
+        return leftPoint;
+    }
+    else if (rightPointInSegment) {
+        return rightPoint;
+    }
+    else {
+        return (bestSegment.first + bestSegment.second)*0.5;
+    }
+
+}
+
+std::pair<Vector2, Vector2> OffensiveCoach::getAimPoints(const Vector2 &fromPoint) {
+    std::pair<Vector2, Vector2> goalSides = world::field->getGoalSides(false);
+    double angleMargin = sin(2.0/180.0*M_PI);
+    double constantMargin = 0.05*world::field->get_field().goal_width;
+    Vector2 leftPoint(goalSides.first.x,
+            goalSides.first.y + constantMargin + angleMargin*goalSides.first.dist(fromPoint));
+    Vector2 rightPoint(goalSides.second.x,
+            goalSides.second.y - angleMargin*goalSides.second.dist(fromPoint) - constantMargin);
+    return std::make_pair(leftPoint, rightPoint);
+}
+
+const std::pair<Vector2, Vector2> &OffensiveCoach::getLongestSegment(
+        const std::vector<std::pair<Vector2, Vector2>> &openSegments) {
+
+    unsigned long bestIndex = 0;
+    for (unsigned long i = 1; i < openSegments.size(); i ++) {
+        auto segment = openSegments[i];
+        auto bestSegment = openSegments[bestIndex];
+        if (abs(segment.first.y - segment.second.y) > abs(bestSegment.first.y - bestSegment.second.y)) {
+            bestIndex = i;
         }
-        else if (rightPointInSegment) {
-            return rightPoint;
+    }
+    return openSegments[bestIndex];
+}
+
+OffensiveCoach::OffensivePosition OffensiveCoach::findBestOffensivePosition(const std::vector<Vector2> &positions,
+        const OffensiveCoach::OffensivePosition &currentBestPosition, const Vector2 &zoneLocation) {
+
+    // get world & field
+    auto world = world::world->getWorld();
+    auto field = world::field->get_field();
+
+    // draw points
```

Comment:
commented code. Make sure atleast the indentation is correct but would prefer to see these removed or to make it an option in the interface to visualize the offense points only when you want to.